File: solutions/114. Flatten Binary Tree to Linked List.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean isleafnode(TreeNode root) {return (root.left == null && root.right == null) ; }
    private boolean hasLeft(TreeNode root) {return (root.left != null) ; }
    
    public void flatten(TreeNode root) {
        if (root == null || isleafnode(root)) {
            return ;
        }
        
        // self
        if (hasLeft(root)) {
            TreeNode rootright = root.right;
            root.right = root.left;
            root.left = null;
            
            TreeNode temp = root.right;
            while (temp.right != null) temp = temp.right;
            temp.right = rootright;
        }
        
        // faith
        flatten(root.right);
        
    }
}

File: solutions/114. Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        
        if not root:
            return
​
        # Flatten left and right subtrees
        self.flatten(root.left)
        self.flatten(root.right)
​
        # Store the right subtree
        right_subtree = root.right
​
        # Move the left subtree to the right
        root.right = root.left
        root.left = None
​
        # Find the tail of the new right subtree
        current = root
        while current.right:
            current = current.right
​
        # Attach the original right subtree
        current.right = right_subtree
        
